server/http/dialog/install.go

```diff
@@ -0,0 +1,150 @@
+		{
+			DisplayName: "App secret:",
+			Name:        "secret",
+			Type:        "text",
+			SubType:     "password",
+			HelpText:    "TODO: How to obtain the App Secret",
+		},
+	}
+	if manifest.RequestedPermissions.Contains(apps.PermissionActAsUser) {
+		elements = append(elements, model.DialogElement{
+			DisplayName: "Require user consent to use REST API first time they use the app:",
+			Name:        "consent",
+			Type:        "radio",
+			Default:     "require",
+			HelpText:    "please indicate if user consent is required to allow the app to act on their behalf",
+			Options: []*model.PostActionOptions{
+				{
+					Text:  "Require user consent",
+					Value: "require",
+				},
+				{
+					Text:  "Do not require user consent",
+					Value: "notrequire",
+				},
+			},
+		})
+	}
+
+	return model.OpenDialogRequest{
+		TriggerId: triggerID,
+		URL:       pluginURL + constants.InteractiveDialogPath + InstallPath,
+		Dialog: model.Dialog{
+			CallbackId:       postID,
+			Title:            "Install App - " + manifest.DisplayName,
+			IntroductionText: intro.String(),
+			Elements:         elements,
+			SubmitLabel:      "Approve and Install",
+			NotifyOnCancel:   true,
+			State:            utils.ToJSON(manifest),
+		},
+	}
+}
+
+func (d *dialog) handleInstall(w http.ResponseWriter, req *http.Request) {
+	var err error
+	rootID := ""
+	message := ""
+	actingUserID := ""
+	status := http.StatusOK
+	defer func() {
+		conf := d.apps.Config.GetConfig()
+		resp := model.SubmitDialogResponse{}
+		if err != nil {
+			resp.Error = errors.Wrap(err, "failed to install").Error()
+			message = "Error: " + resp.Error
+		}
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(status)
+		_ = json.NewEncoder(w).Encode(resp)
+
+		if actingUserID != "" {
+			_ = d.apps.Mattermost.Post.DM(conf.BotUserID, actingUserID, &model.Post{
+				RootId:   rootID,
+				ParentId: rootID,
+				Message:  message,
+			})
+		}
+	}()
+
+	actingUserID = req.Header.Get("Mattermost-User-Id")
+	if actingUserID == "" {
+		err = errors.New("user not logged in")
+		status = http.StatusUnauthorized
+		return
+	}
+	// <><> TODO check for sysadmin
+
+	var dialogRequest model.SubmitDialogRequest
+	err = json.NewDecoder(req.Body).Decode(&dialogRequest)
+	if err != nil {
+		status = http.StatusBadRequest
+		return
+	}
+	if dialogRequest.Type != "dialog_submission" {
+		err = errors.New("expected dialog_submission, got " + dialogRequest.Type)
+		status = http.StatusBadRequest
+		return
+	}
+
+	rootID = dialogRequest.CallbackId
+
+	if dialogRequest.Cancelled {
+		message = "Installation was canceled by the user"
+		return
+	}
+
+	v := dialogRequest.Submission["consent"]
+	consentValue, _ := v.(string)
+	noUserConsentForOAuth2 := false
+	if consentValue == "require" {
+		noUserConsentForOAuth2 = true
+	}
+
+	v = dialogRequest.Submission["secret"]
+	secret, _ := v.(string)
+
+	var manifest apps.Manifest
+	err = json.Unmarshal([]byte(dialogRequest.State), &manifest)
```

Comment:
We are using the manifest that comes with the request. This might be a security issue if someone was able to change the content of the manifest on the client side, then making the user accept some permissions while the underlying manifest has others.

In order to solve this, in the original PR I did some countermeasures inspired by OAuth:
- I create a unique ID for the dialog, and create a "state" variable in the dialog.
- I add a register in the KVStore with the ID as key, and the manifest as value
- When the dialog returns, I check the state, and check if that state is in the KVStore.
- If it is in the KVStore, I authorize the manifest I have in the store, since it is the one I used to create the dialog

Probably it is not a perfect solution, and I may be making things more secure than they need, but I thought it was a nice addition.

The relevant code is here:
https://github.com/mattermost/mattermost-plugin-cloudapps/blob/05a96c274a064f0ca66701b1f64f533908282500/server/authorization.go#L55-L67
https://github.com/mattermost/mattermost-plugin-cloudapps/blob/05a96c274a064f0ca66701b1f64f533908282500/server/authorization.go#L81-L86
https://github.com/mattermost/mattermost-plugin-cloudapps/blob/05a96c274a064f0ca66701b1f64f533908282500/server/authorization.go#L131-L150

Reply:
If a remote user session is compromised, it can just execute necessary APIs to install whatever it wants, I am kinda puzzled with the need for the extra verification steps here.

0/5 we mark this (obviously! regardless of how we verify the steps) for a review with the security team once all of the Auth flows are in place?